File: backend/src/adapters/stripe.ts
import Stripe from 'stripe';
import {
  CreatePaymentRequest,
  CreatePaymentResponse,
  RefundPaymentResponse,
  ListPaymentsParams,
  AdapterListPaymentsResult,
  PaymentRecord,
  PaymentStatus,
} from '../types/payment.js';
import { PaymentAdapter } from './base.js';
import { auditLog, errorLog } from '../utils/logger.js';
import supabase from '../utils/supabase.js';

export class StripeAdapter implements PaymentAdapter {
  private stripe: Stripe;

  constructor(apiKey?: string) {
    const key = apiKey || process.env.STRIPE_API_KEY;
    if (!key) {
      throw new Error('Stripe API key not provided');
    }
    this.stripe = new Stripe(key, {
      apiVersion: '2024-06-20',
    });
  }

  async createPayment(request: CreatePaymentRequest): Promise<CreatePaymentResponse> {
    try {
      const paymentIntent = await this.stripe.paymentIntents.create({
        amount: toStripeAmount(request.amount),
        currency: request.currency.toLowerCase(),
        payment_method: request.payment_method,
        confirm: true,
        automatic_payment_methods: {
          enabled: true,
          allow_redirects: 'never',
        },
        description: request.description,
        metadata: {
          customer_id: request.customer_id,
          ...request.metadata,
        },
      });

      auditLog('PAYMENT_CREATED', {
        provider: 'stripe',
        transaction_id: paymentIntent.id,
        amount: request.amount,
        currency: request.currency,
        customer_id: request.customer_id,
      });

      return {
        id: paymentIntent.id,
        provider_transaction_id: paymentIntent.id,
        amount: request.amount,
        currency: paymentIntent.currency?.toUpperCase() || request.currency,
        status: mapStripeStatus(paymentIntent.status),
        created_at: new Date(paymentIntent.created * 1000).toISOString(),
        metadata: request.metadata,
      };
    } catch (error) {
      errorLog(error, { context: 'Stripe payment creation failed', request });
      throw error;
    }
  }

  async refundPayment(
    transactionId: string,
    amount?: number,
    reason?: string
  ): Promise<RefundPaymentResponse> {
    try {
      const refund = await this.stripe.refunds.create({
        payment_intent: transactionId,
        amount: amount ? toStripeAmount(amount) : undefined,
        reason: reason as Stripe.RefundCreateParams.Reason,
      });

      auditLog('PAYMENT_REFUNDED', {
        provider: 'stripe',
        original_transaction_id: transactionId,
        refund_id: refund.id,
        amount: refund.amount ? fromStripeAmount(refund.amount) : amount,
      });

      return {
        refund_id: refund.id,
        original_transaction_id: transactionId,
        amount: refund.amount ? fromStripeAmount(refund.amount) : amount || 0,
        status: mapStripeStatus(refund.status),
        created_at: new Date(refund.created * 1000).toISOString(),
      };
    } catch (error) {
      errorLog(error, { context: 'Stripe refund failed', transactionId });
      throw error;
    }
  }

  async listPayments(params: ListPaymentsParams): Promise<AdapterListPaymentsResult> {
    const limit = params.limit ?? 10;
    const offset = params.offset ?? 0;

    let query = supabase
      .from('payments')
      .select('*', { count: 'exact' })
      .eq('provider', 'stripe')
      .order('created_at', { ascending: false })
      .range(offset, offset + limit - 1);

    if (params.status) {
      query = query.eq('status', params.status);
    }
    if (params.customer_id) {
      query = query.eq('customer_id', params.customer_id);
    }
    if (params.start_date) {
      query = query.gte('created_at', params.start_date);
    }
    if (params.end_date) {
      query = query.lte('created_at', params.end_date);
    }

    const { data, error, count } = await query;
    if (error) {
      errorLog(error, { context: 'Stripe adapter listPayments failed' });
      throw error;
    }

    const payments: PaymentRecord[] = (data || []).map((record) => this.mapSupabaseRecord(record));

    return {
      payments,
      total: count ?? payments.length,
    };
  }

  private mapSupabaseRecord(record: Record<string, unknown>): PaymentRecord {
    return {
      id: String(record.id),
      provider_transaction_id: String(record.provider_transaction_id),
      provider: 'stripe',
      amount: Number(record.amount),
      currency: String(record.currency).toUpperCase(),
      status: (record.status as PaymentStatus) || 'pending',
      customer_id: (record.customer_id as string) ?? null,
      metadata: (record.metadata as Record<string, unknown>) ?? null,
      refund_id: (record.refund_id as string) ?? null,
      refund_status: (record.refund_status as string) ?? null,
      refund_amount:
        record.refund_amount === null || record.refund_amount === undefined
          ? null
          : Number(record.refund_amount),
      created_at: String(record.created_at),
      updated_at: String(record.updated_at),
    };
  }
}

const STRIPE_STATUS_MAP: Record<string, PaymentStatus> = {
  succeeded: 'completed',
  requires_action: 'processing',
  requires_capture: 'processing',
  requires_confirmation: 'processing',
  requires_payment_method: 'failed',
  processing: 'processing',
  canceled: 'failed',
  failed: 'failed',
  pending: 'pending',
  refunded: 'refunded',
};

const toStripeAmount = (amount: number): number => Math.round(amount * 100);

const fromStripeAmount = (amount: number): number => Number((amount / 100).toFixed(2));

const mapStripeStatus = (status?: string | null): PaymentStatus => {
  if (!status) {
    return 'pending';
  }

  return STRIPE_STATUS_MAP[status] || 'pending';
};

File: backend/app/payments/providers/stripe.py
"""Stripe payment provider adapter.

Implements the PaymentProviderAdapter interface for Stripe payments.
Uses the official Stripe Python SDK.
"""

from __future__ import annotations

from typing import Any

import stripe
from stripe import PaymentIntent, Refund
from stripe._error import InvalidRequestError, StripeError

from app.config import Settings
from app.logging import get_logger
from app.payments.errors import (
    PaymentFailedError,
    PaymentNotFoundError,
    ProviderError,
    RefundFailedError,
)
from app.payments.providers.base import (
    PaymentProviderAdapter,
    ProviderPaymentResult,
    ProviderRefundResult,
    ProviderStatusResult,
)
from app.payments.types import PaymentStatus

logger = get_logger("payments.stripe")


class StripeAdapter(PaymentProviderAdapter):
    """Stripe payment provider adapter.

    Handles payment creation, refunds, and status checks via Stripe API.
    """

    def __init__(self, settings: Settings) -> None:
        """Initialize the Stripe adapter.

        Args:
            settings: Application settings containing Stripe API key.

        Raises:
            ValueError: If Stripe API key is not configured.
        """
        if not settings.stripe_api_key:
            raise ValueError("STRIPE_API_KEY is not configured")

        self.api_key = settings.stripe_api_key
        stripe.api_key = self.api_key
        stripe.api_version = "2024-11-20.acacia"

    @property
    def provider_name(self) -> str:
        """Return the provider name."""
        return "stripe"

    def _map_stripe_status(self, stripe_status: str) -> PaymentStatus:
        """Map Stripe PaymentIntent status to our PaymentStatus.

        Args:
            stripe_status: Stripe's payment intent status.

        Returns:
            Mapped PaymentStatus.
        """
        status_map = {
            "requires_payment_method": PaymentStatus.PENDING,
            "requires_confirmation": PaymentStatus.PENDING,
            "requires_action": PaymentStatus.PENDING,
            "processing": PaymentStatus.PROCESSING,
            "requires_capture": PaymentStatus.PROCESSING,
            "canceled": PaymentStatus.FAILED,
            "succeeded": PaymentStatus.COMPLETED,
        }
        return status_map.get(stripe_status, PaymentStatus.PENDING)

    def _map_refund_status(self, refund_status: str) -> PaymentStatus:
        """Map Stripe Refund status to our PaymentStatus.

        Args:
            refund_status: Stripe's refund status.

        Returns:
            Mapped PaymentStatus.
        """
        status_map = {
            "succeeded": PaymentStatus.REFUNDED,
            "pending": PaymentStatus.PROCESSING,
            "failed": PaymentStatus.FAILED,
            "canceled": PaymentStatus.FAILED,
            "requires_action": PaymentStatus.PROCESSING,
        }
        return status_map.get(refund_status, PaymentStatus.PROCESSING)

    async def create_payment(
        self,
        amount: int,
        currency: str,
        payment_method: str,
        customer_id: str,
        description: str | None = None,
        metadata: dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> ProviderPaymentResult:
        """Create a payment via Stripe PaymentIntent.

        Creates a PaymentIntent and optionally confirms it if payment_method
        starts with 'pm_' (indicating a saved payment method).

        Args:
            amount: Amount in smallest currency unit.
            currency: Three-letter ISO currency code.
            payment_method: Stripe payment method ID or token.
            customer_id: Customer identifier for metadata.
            description: Optional payment description.
            metadata: Optional metadata.
            idempotency_key: Optional idempotency key.

        Returns:
            ProviderPaymentResult with transaction details.

        Raises:
            PaymentFailedError: If the payment fails.
            ProviderError: If there's a Stripe API error.
        """
        normalized_currency = self.normalize_currency(currency)

        try:
            logger.info(
                "Creating Stripe PaymentIntent",
                amount=amount,
                currency=normalized_currency,
                payment_method=payment_method[:20] + "..."
                if len(payment_method) > 20
                else payment_method,
            )

            # Build metadata for Stripe
            stripe_metadata: dict[str, str] = {
                "customer_id": customer_id,
            }
            if metadata:
                # Stripe metadata must be strings
                for key, value in metadata.items():
                    stripe_metadata[key] = str(value)

            # Create PaymentIntent params
            intent_params: dict[str, Any] = {
                "amount": amount,
                "currency": normalized_currency.lower(),
                "payment_method": payment_method,
                "metadata": stripe_metadata,
                "confirm": True,  # Auto-confirm for server-side payments
                "automatic_payment_methods": {
                    "enabled": True,
                    "allow_redirects": "never",
                },
            }

            if description:
                intent_params["description"] = description

            # Add idempotency key if provided
            request_options: dict[str, Any] = {}
            if idempotency_key:
                request_options["idempotency_key"] = idempotency_key

            # Create the PaymentIntent
            payment_intent: PaymentIntent = stripe.PaymentIntent.create(
                **intent_params,
                **request_options,
            )

            logger.info(
                "Stripe PaymentIntent created",
                payment_intent_id=payment_intent.id,
                status=payment_intent.status,
            )

            return ProviderPaymentResult(
                provider_transaction_id=payment_intent.id,
                status=self._map_stripe_status(payment_intent.status),
                client_secret=payment_intent.client_secret,
                provider_metadata={
                    "stripe_status": payment_intent.status,
                    "payment_method_types": payment_intent.payment_method_types,
                    "livemode": payment_intent.livemode,
                },
            )

        except InvalidRequestError as e:
            logger.error(
                "Stripe invalid request error",
                error=str(e),
                code=e.code,
                param=e.param,
            )
            raise PaymentFailedError(
                message=f"Payment failed: {e.user_message or str(e)}",
                details={
                    "provider": "stripe",
                    "stripe_code": e.code,
                    "stripe_param": e.param,
                },
            )
        except StripeError as e:
            logger.error(
                "Stripe API error",
                error=str(e),
                error_type=type(e).__name__,
            )
            raise ProviderError(
                provider="stripe",
                message=f"Stripe error: {e.user_message or str(e)}",
                details={
                    "stripe_error_type": type(e).__name__,
                },
            )

    async def refund_payment(
        self,
        provider_transaction_id: str,
        amount: int | None = None,
        reason: str | None = None,
        idempotency_key: str | None = None,
    ) -> ProviderRefundResult:
        """Refund a Stripe payment.

        Args:
            provider_transaction_id: Stripe PaymentIntent ID.
            amount: Optional partial refund amount. If None, full refund.
            reason: Optional refund reason.
            idempotency_key: Optional idempotency key.

        Returns:
            ProviderRefundResult with refund details.

        Raises:
            RefundFailedError: If the refund fails.
            PaymentNotFoundError: If the PaymentIntent doesn't exist.
            ProviderError: If there's a Stripe API error.
        """
        try:
            logger.info(
                "Creating Stripe refund",
                payment_intent=provider_transaction_id,
                amount=amount,
            )

            refund_params: dict[str, Any] = {
                "payment_intent": provider_transaction_id,
            }

            if amount is not None:
                refund_params["amount"] = amount

            if reason:
                # Stripe accepts: duplicate, fraudulent, requested_by_customer
                stripe_reason = "requested_by_customer"
                if "duplicate" in reason.lower():
                    stripe_reason = "duplicate"
                elif "fraud" in reason.lower():
                    stripe_reason = "fraudulent"
                refund_params["reason"] = stripe_reason
                refund_params["metadata"] = {"original_reason": reason}

            # Add idempotency key if provided
            request_options: dict[str, Any] = {}
            if idempotency_key:
                request_options["idempotency_key"] = idempotency_key

            refund: Refund = stripe.Refund.create(
                **refund_params,
                **request_options,
            )

            logger.info(
                "Stripe refund created",
                refund_id=refund.id,
                status=refund.status,
                amount=refund.amount,
            )

            return ProviderRefundResult(
                refund_id=refund.id,
                status=self._map_refund_status(refund.status),
                amount=refund.amount,
                provider_metadata={
                    "stripe_status": refund.status,
                    "stripe_reason": refund.reason,
                    "livemode": refund.livemode,
                },
            )

        except InvalidRequestError as e:
            logger.error(
                "Stripe refund invalid request",
                error=str(e),
                code=e.code,
            )
            # Check if payment not found
            if e.code == "resource_missing":
                raise PaymentNotFoundError(
                    payment_id=provider_transaction_id,
                    details={"provider": "stripe"},
                )
            raise RefundFailedError(
                message=f"Refund failed: {e.user_message or str(e)}",
                details={
                    "provider": "stripe",
                    "stripe_code": e.code,
                },
            )
        except StripeError as e:
            logger.error(
                "Stripe refund API error",
                error=str(e),
            )
            raise ProviderError(
                provider="stripe",
                message=f"Stripe refund error: {e.user_message or str(e)}",
            )

    async def get_payment_status(
        self,
        provider_transaction_id: str,
    ) -> ProviderStatusResult:
        """Get the current status of a Stripe payment.

        Args:
            provider_transaction_id: Stripe PaymentIntent ID.

        Returns:
            ProviderStatusResult with current status.

        Raises:
            PaymentNotFoundError: If the PaymentIntent doesn't exist.
            ProviderError: If there's a Stripe API error.
        """
        try:
            logger.info(
                "Fetching Stripe PaymentIntent status",
                payment_intent=provider_transaction_id,
            )

            payment_intent = stripe.PaymentIntent.retrieve(provider_transaction_id)

            # Check for refunds
            status = self._map_stripe_status(payment_intent.status)

            # If there are charges, check if any are refunded
            if payment_intent.latest_charge:
                charge = stripe.Charge.retrieve(payment_intent.latest_charge)
                if charge.refunded:
                    status = PaymentStatus.REFUNDED

            return ProviderStatusResult(
                status=status,
                provider_metadata={
                    "stripe_status": payment_intent.status,
                    "amount_received": payment_intent.amount_received,
                    "livemode": payment_intent.livemode,
                },
            )

        except InvalidRequestError as e:
            if e.code == "resource_missing":
                raise PaymentNotFoundError(
                    payment_id=provider_transaction_id,
                    details={"provider": "stripe"},
                )
            raise ProviderError(
                provider="stripe",
                message=f"Stripe error: {e.user_message or str(e)}",
            )
        except StripeError as e:
            raise ProviderError(
                provider="stripe",
                message=f"Stripe error: {e.user_message or str(e)}",
            )
